Implement bare-metal Mac M1 podman-machine testing

cevich opened the pull request:
***Depends on***: https://github.com/containers/podman/pull/20852

Setup and execute podman machine testing on bare-metal M1 Macs
using a pool of shared and semi-persistent hosts.  Automated
and manual processes outside this repository are responsible
for providing and maintaining all hosts.  Ref.
https://github.com/containers/automation/tree/main/mac_pw_pool

Update the `localmachine` make target to standardize execution
across platforms.  Update/simplify podman-machine e2e README to
reflect current reality.

Warning: This CI setup and supporting infrastructure was developed
in favor of expediency vs reliability and stability.  There are
many possible failure-modes (known and unknown) which may lead
to undefined test behaviors.  Future work may address some of
these as they are encountered or discovered.

[NO NEW TESTS NEEDED]

#### Does this PR introduce a user-facing change?

```release-note
None
```

cevich commented:
This is ready for initial review.

cevich commented:
Review note:  The basic tests all pass, however the full suite fails quite badly.  I'm intending to capture the failure details into a new issue before disabling the full-suite with a FIXME.

edsantiago commented:
Task map. New task is the bottommost cyan box near bottom right ![cirrus-map-pr20691](https://github.com/containers/podman/assets/265639/dd990c1e-55b5-424b-a8ef-2e7067698bce)

cevich commented:
Staged changes:

* A rebase from 54425af0a610b9d86ee33339914611a51e6391cd w/ added dependency on https://github.com/containers/podman/pull/20852:
* Added commit: `Do not aggregate failing mac test status`.  Made note of this in issue https://github.com/containers/podman/issues/20853
* Fixed `^M` whitespace in README.
* Removed test runner script, simplified into into two separate invocations.  One for basic, followed by a second for the full suite.

edsantiago commented on Makefile:
```diff
@@ -624,8 +624,8 @@ localintegration: test-binaries ginkgo
 .PHONY: localmachine
-localmachine: test-binaries .install.ginkgo
-	$(MAKE) ginkgo-run GINKGO_PARALLEL=n GINKGOWHAT=pkg/machine/e2e/. HACK=
+localmachine:
+	$(MAKE) ginkgo-run GINKGO_PARALLEL=n TAGS="$(REMOTETAGS)" GINKGO_FLAKE_ATTEMPTS=0 FOCUS_FILE=$(FOCUS_FILE) GINKGOWHAT=pkg/machine/e2e/. HACK=
```
This will be removed before commit, right?

cevich replied:
There's a dedicated commit for this that explains why it's needed.  I checked carefully to make sure it /should/ work for all podman-machine CI tasks (except windows, it uses `winmake`).  If we quay would cooperate, we could actually verify that's the case :confounded: 

Or...It's not obvious to my overstressed-brain, could you please expand on what you're concerned about?

cevich replied:
If it helps, some more context:

* The machine tests already take a long time to run, I don't think we want any automatic flake re-runs.
* For machine tests, the tag-list should always be `REMOTETAGS`.
* The `FOCUS_FILE` value is empty by default (to run all tests), it's needed for the new `smoke_test` script.
* `test-binaries` fails on Mac likely b/c complex linux-only podman (not podman-remote) reasons (I don't remember the specifics).  In any case it's unnecessary for this target b/c podman/podman-remote targets require it as needed, and `localmachine` fails w/o pre-built binaries.
* `.install.ginkgo` is unnecessary b/c `ginkgo-run` depends on it already.

edsantiago replied:
Thanks for the clarification. My bad, I didn't chase those down deep enough.

edsantiago commented on contrib/cirrus/mac_cleanup.sh:
```diff
@@ -0,0 +1,25 @@
+# Since it run both before and after tasks, it must exit cleanly if there was a cleanup
+# failure (i.e. file or directory not found).
+
+set +ex
```
I'd much rather `+e -x` for something that will need debugging one day

cevich replied:
Oops, my bad.  I always get confused about `-x` and `+x`.  You are correct, I was intending `-x` to enable debugging.  Good catch!

cevich replied:
Fix staged.

edsantiago commented on contrib/cirrus/mac_env.sh:
```diff
@@ -0,0 +1,25 @@
+
+# Many variables can affect operations, make them all known to assist debugging.
+echo "Selection of current env. vars:"
+for env_var_name in $(awk 'BEGIN{for(v in ENVIRON) print v}' | grep -Eiv '(^PATH$)|(^BASH_FUNC)|(^_.*)' | sort -u)
```
I would hope that `-u` will never be needed (in `sort -u`) for envariables?!? As a casual reader this cost me time to figure out why it's there. Could you remove it to save someone else that time?

edsantiago commented on contrib/cirrus/mac_env.sh:
```diff
@@ -0,0 +1,25 @@
+
+set -euo pipefail
+
+# This script is intended to be called by Cirrus-CI on a Mac M1 persistent workier.
```
Typo, 'workier'.

Also, in each header for each of these scripts, would you consider adding something like "this is step X of 3" and/or "this runs before/after X", so someone looking in the overflowing `contrib/cirrus ` directory can more easily figure out the order in which these run?

cevich replied:
Got it.  Sure thing, will do.  Fixes staged.

edsantiago reviewed (commented):
Only partway through but submitting before github clobbers my review comments

edsantiago commented on .cirrus.yml:
```diff
@@ -796,6 +801,51 @@ podman_machine_windows_task:
+    alias: podman_machine_mac
+    only_if: *no_rhel_release
+    depends_on:
+      - build_aarch64
```
This makes for a weird-looking graph. How necessary is it really?

cevich replied:
Hmmm, ahh yes, it only needs to depend on `osx_alt_build`.  Fix staged.

cevich commented on contrib/cirrus/mac_setup.sh:
```diff
@@ -0,0 +1,30 @@
+
+set -euo pipefail
+
+# Persist env. vars required during testing steps.  These cannot be set in the
+# task YAML because they would interfere with repo. cloning and task preperation.
+# Note: None of the values below will become active until the next "script" section starts.
+# Ref:
+# https://cirrus-ci.org/guide/tips-and-tricks/#setting-environment-variables-from-scripts
```
Sure.  I'll also note it's a substitution for `/etc/ci_environment` because these instances are shared for multiple tasks.  Fix staged.

cevich commented:
Force push: Incremental fixes to address Ed's comments as of 2f4095529ea12aeb9ec00b2b447a3c9acf7a7391

cevich commented:
force-push: conflict resolution, no code changes.

edsantiago commented:
LGTM with fingers crossed...

cevich commented:
Thanks for all your help Ed.  There are going to be hiccups - this being so drastically orthogonal to how all other CI works.  I think I have enough monitoring in place to help debug instance problems, if not I'll work on adding more as needed.

cevich commented:
@ashley-cui @baude @mheon PTAL :crossed_fingers:

edsantiago commented:
FWIW I restarted the machine-mac test once already, it [timed out](https://api.cirrus-ci.com/v1/task/4804740878434304/logs/test.log)

cevich commented:
**important note**: The `timeout` failure of the main tests in `machine-mac podman darwin rootless host sqlite` is completely expected.  At this stage, the only thing that should be green is `smoke_test`.

cevich commented:
Restarted [`sys podman fedora-38 root host boltdb` again (failure log)](https://api.cirrus-ci.com/v1/artifact/task/4815557485133824/html/sys-podman-fedora-38-root-host-boltdb.log.html).  Didn't look deeply, but the problem appears not good:

```
<+031ms> # # podman run -d --name=test-ctr --rm quay.io/libpod/testimage:20221018 sh -c mkdir -p /foo/test1. /foo/test2; touch /foo/test1./file1 /foo/test2/file2; sleep infinity
<+382ms> # 47d87113ab959ae2ac6019568a26f2f77e9b2e0a3938babb873f0dfcf5ba0413
         #
<+011ms> # # podman cp test-ctr:/foo/test1. /var/tmp/podman_bats.jfntbd/dst/foo
<+063ms> # Error: "/foo/test1." could not be found on container test-ctr: no such file or directory
<+006ms> # [ rc=125 (** EXPECTED 0 **) ]
```

Maybe the `run -d` is racing?

cevich commented:
Re-ran `sys podman fedora-38 root host boltdb` again.  Different test-failure this time :confused:

cevich commented:
All green now

edsantiago commented:
Here goes...

/lgtm

edsantiago commented:
/approve
